In [1]:
import pandas as pd
import psycopg2 as pg2
from sqlalchemy import create_engine

engine = create_engine('postgresql://testuser:testpass@localhost:5432/postgresql_analysis')

con = pg2.connect(host='localhost',
                  user='testuser',
                  password='testpass',
                  database='postgresql_analysis')
con.autocommit = True
cur = con.cursor()

In [2]:
def select(query):
    return pd.read_sql(query, con)

In [3]:
query = """
        SELECT * FROM orders
        """

select(query)

order_id                       customer_id  \
0      e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1      53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
2      47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
3      949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   
4      ad21c59c0840e6cb83a9ceb5573f8159  8ab97904e6daea8866dbdbc4fb7aad2c   
...                                 ...                               ...   
99436  9c5dedf39a927c1b2549525ed64a053c  39bd1228ee8140590ac3aca26f2dfe00   
99437  63943bddc261676b46f01ca7ac2f7bd8  1fca14ff2861355f6e5f14306ff977a7   
99438  83c1379a015df1e13d02aae0204711ab  1aa71eb042121263aafbe80c1b562c9c   
99439  11c177c8e97725db2631073c19f07b62  b331b74b18dc79bcdf6532d51e1637c1   
99440  66dea50a8b16d9b4dee7af250b4be1a5  edb027a75a1449115f6b43211ae02a24   

      order_status order_purchase_timestamp   order_approved_at  \
0        delivered      2017-10-02 10:56:33 2017-10-02 11:07:15   
1        delivered      2018-07-24 20:41:37 2018-07-26 03:24:27   
2        delivered      2018-08-08 08:38:49 2018-08-08 08:55:23   
3        delivered      2017-11-18 19:28:06 2017-11-18 19:45:59   
4        delivered      2018-02-13 21:18:39 2018-02-13 22:20:29   
...            ...                      ...                 ...   
99436    delivered      2017-03-09 09:54:05 2017-03-09 09:54:05   
99437    delivered      2018-02-06 12:58:58 2018-02-06 13:10:37   
99438    delivered      2017-08-27 14:46:43 2017-08-27 15:04:16   
99439    delivered      2018-01-08 21:28:27 2018-01-08 21:36:21   
99440    delivered      2018-03-08 20:57:30 2018-03-09 11:20:28   

      order_delivered_carrier_date order_delivered_customer_date  \
0              2017-10-04 19:55:00           2017-10-10 21:25:13   
1              2018-07-26 14:31:00           2018-08-07 15:27:45   
2              2018-08-08 13:50:00           2018-08-17 18:06:29   
3              2017-11-22 13:39:59           2017-12-02 00:28:42   
4              2018-02-14 19:46:34           2018-02-16 18:17:02   
...                            ...                           ...   
99436          2017-03-10 11:18:03           2017-03-17 15:08:01   
99437          2018-02-07 23:22:42           2018-02-28 17:37:56   
99438          2017-08-28 20:52:26           2017-09-21 11:24:17   
99439          2018-01-12 15:35:03           2018-01-25 23:32:54   
99440          2018-03-09 22:11:59           2018-03-16 13:08:30   

      order_estimated_delivery_date  
0                        2017-10-18  
1                        2018-08-13  
2                        2018-09-04  
3                        2017-12-15  
4                        2018-02-26  
...                             ...  
99436                    2017-03-28  
99437                    2018-03-02  
99438                    2017-09-27  
99439                    2018-02-15  
99440                    2018-04-03  

[99441 rows x 8 columns]

In [4]:
query = """
        SELECT * FROM order_items
        """

select(query)

order_id order_item_id  \
0       00010242fe8c5a6d1ba2dd792cb16214             1   
1       00018f77f2f0320c557190d7a144bdd3             1   
2       000229ec398224ef6ca0657da4fc703e             1   
3       00024acbcdf0a6daa1e931b038114c75             1   
4       00042b26cf59d7ce69dfabb4e55b4fd9             1   
...                                  ...           ...   
112645  fffc94f6ce00a00581880bf54a75a037             1   
112646  fffcd46ef2263f404302a634eb57f7eb             1   
112647  fffce4705a9662cd70adb13d4a31832d             1   
112648  fffe18544ffabc95dfada21779c9644f             1   
112649  fffe41c64501cc87c801fd61db3f6244             1   

                              product_id                         seller_id  \
0       4244733e06e7ecb4970a6e2683c13e61  48436dade18ac8b2bce089ec2a041202   
1       e5f2d52b802189ee658865ca93d83a8f  dd7ddc04e1b6c2c614352b383efe2d36   
2       c777355d18b72b67abbeef9df44fd0fd  5b51032eddd242adc84c38acab88f23d   
3       7634da152a4610f1595efa32f14722fc  9d7a1d34a5052409006425275ba1c2b4   
4       ac6c3623068f30de03045865e4e10089  df560393f3a51e74553ab94004ba5c87   
...                                  ...                               ...   
112645  4aa6014eceb682077f9dc4bffebc05b0  b8bc237ba3788b23da09c0f1f3a3288c   
112646  32e07fd915822b0765e448c4dd74c828  f3c38ab652836d21de61fb8314b69182   
112647  72a30483855e2eafc67aee5dc2560482  c3cfdc648177fdbbbb35635a37472c53   
112648  9c422a519119dcad7575db5af1ba540e  2b3e4a2a3ea8e01938cabda2a3e5cc79   
112649  350688d9dc1e75ff97be326363655e01  f7ccf836d21b2fb1de37564105216cc1   

       shipping_limit_date   price  freight_value  
0      2017-09-19 09:45:35   58.90          13.29  
1      2017-05-03 11:05:13  239.90          19.93  
2      2018-01-18 14:48:30  199.00          17.87  
3      2018-08-15 10:10:18   12.99          12.79  
4      2017-02-13 13:57:51  199.90          18.14  
...                    ...     ...            ...  
112645 2018-05-02 04:11:01  299.99          43.41  
112646 2018-07-20 04:31:48  350.00          36.53  
112647 2017-10-30 17:14:25   99.90          16.95  
112648 2017-08-21 00:04:32   55.99           8.72  
112649 2018-06-12 17:10:13   43.00          12.79  

[112650 rows x 7 columns]

### Check order_status categories

In [123]:
query = """
        SELECT a.*
            ,num_orders * 1.0 / count(o.order_id) * 100 as percentage
        FROM 
        (
            SELECT order_status, count(order_id) as num_orders
            FROM orders
            GROUP BY order_status
        ) a
        CROSS JOIN orders o
        GROUP BY a.order_status, num_orders
        ORDER BY num_orders desc
        """

select(query)

order_status  num_orders  percentage
0    delivered       96478   97.020344
1      shipped        1107    1.113223
2     canceled         625    0.628513
3  unavailable         609    0.612423
4     invoiced         314    0.315765
5   processing         301    0.302692
6      created           5    0.005028
7     approved           2    0.002011

97% orders are delivered so, we will focus on delivered orders only.

### On-time delivery rate

Olist requires merchants to maintain their on-time delivery rate at least 97% for the last 30 days to guarantee their performance.

I would like to know the number of sellers who have 97% or higher on-time delivery rate lately. First, let's check if there are any sellers who delivered orders on time each year in 2018.

In [99]:
query = """
        SELECT date_part('month', order_purchase_timestamp)::int as month,
               count(distinct seller_id) as num_sellers
        FROM 
        (
            SELECT oi.seller_id, o.order_purchase_timestamp
            FROM orders o 
            JOIN order_items oi on o.order_id = oi.order_id
            WHERE order_status = 'delivered'
            and date_part('year', order_purchase_timestamp) = 2018
            and date_part('day', age(o.order_estimated_delivery_date, o.order_delivered_customer_date)) >= 0
        ) a
        GROUP BY month
        """

select(query)

month  num_sellers
0      1          933
1      2          878
2      3          923
3      4         1084
4      5         1085
5      6         1158
6      7         1225
7      8         1234

The nearest month we have is August. The query below calculate the number of sellers who have 97% or higher on-time delivery rate in August 2018.

In [153]:
query = """
        SELECT c.num_sellers, count(b.seller_id) as num_qualified_sellers
               ,count(b.seller_id) * 1.0 / c.num_sellers * 100 as percentage
        FROM 
        (
            SELECT a.seller_id
                ,num_on_time_orders * 1.0 / count(distinct order_id) * 100 as percent_on_time
            FROM order_items
            JOIN 
            (
                SELECT oi.seller_id
                    ,count(distinct o.order_id) as num_on_time_orders
                FROM orders o 
                JOIN order_items oi on o.order_id = oi.order_id
                WHERE order_status = 'delivered'
                and date_part('month', order_purchase_timestamp) = 8
                and date_part('year', order_purchase_timestamp) = 2018
                and date_part('day', age(o.order_estimated_delivery_date, o.order_delivered_customer_date)) >= 0
                GROUP BY 1
            ) a
            on order_items.seller_id = a.seller_id
            GROUP BY a.seller_id, num_on_time_orders
            HAVING num_on_time_orders * 1.0 / count(distinct order_id) * 100 >= 97
        ) b
        CROSS JOIN 
        (
            SELECT count(distinct seller_id) as num_sellers
            FROM orders o 
            JOIN order_items oi on o.order_id = oi.order_id
            WHERE order_status = 'delivered'
            and date_part('month', order_purchase_timestamp) = 8
            and date_part('year', order_purchase_timestamp) = 2018
        ) c
        GROUP BY 1
        """

select(query)

num_sellers  num_qualified_sellers  percentage
0         1261                    135   10.705789

In August 2018, among 1261 sellers who have delivered orders to customers, only 135 (10.71%) have 97% on-time delivery rate.

In [154]:
query = """
        SELECT a.seller_id
            ,count(distinct order_id) as num_delivered_orders
            ,num_on_time_orders
            ,num_on_time_orders * 1.0 / count(distinct order_id) * 100 as percent_on_time
        FROM order_items
        JOIN
        (
            SELECT oi.seller_id
                ,count(distinct o.order_id) as num_on_time_orders
            FROM orders o 
            JOIN order_items oi on o.order_id = oi.order_id
            WHERE order_status = 'delivered'
            and date_part('month', order_purchase_timestamp) = 8
            and date_part('year', order_purchase_timestamp) = 2018
            and date_part('day', age(o.order_estimated_delivery_date, o.order_delivered_customer_date)) >= 0
            GROUP BY 1
        ) a
        on order_items.seller_id = a.seller_id
        GROUP BY 1,3
        HAVING num_on_time_orders * 1.0 / count(distinct order_id) * 100 >= 97
        ORDER BY 2 desc
        LIMIT 10
        """

select(query)

seller_id  num_delivered_orders  num_on_time_orders  \
0  81f89e42267213cb94da7ddc301651da                    46                  46   
1  aac51c486b672a9850d59f3e84b1cf88                    10                  10   
2  402916f742e5c740cc751493d9cf5053                     9                   9   
3  55dedd83e501d8248880557d9073cbfd                     8                   8   
4  ff1e15b778c700abdd4d239b81ac466d                     7                   7   
5  c3aad7dc65449ae90a5e9c3c6c1e78e0                     6                   6   
6  7f40d06aa0b5f1aa4f41af8c0480e2ef                     6                   6   
7  c5e60e39c0f42b8e827daa13cff74afa                     6                   6   
8  09bad886111255c5b5030314fc7f1a4a                     5                   5   
9  7b7ed56ae755e2880f2b69632e839c1e                     5                   5   

   percent_on_time  
0            100.0  
1            100.0  
2            100.0  
3            100.0  
4            100.0  
5            100.0  
6            100.0  
7            100.0  
8            100.0  
9            100.0

Here is the top 10 sellers who have the highest number of delivered orders with 97% on-time delivery rate in August 2018.

### How fast is it for a customer to receive an order?

This query calculates the percentage of orders arriving within 2 days, 1 week, 2 weeks, or more than 2 weeks after they are placed.

In [163]:
query = """
        SELECT
            sum(case
                when date_part('day', 
                age(order_delivered_customer_date, 
                    order_purchase_timestamp)) <= 2 then 1 
                end) * 1.0 / count(order_id) * 100 as under_two_days,
            sum(case when date_part('day',
                age(order_delivered_customer_date, 
                    order_purchase_timestamp)) BETWEEN 3 and 5 then 1
                end) * 1.0 / count(order_id) * 100 as in_one_week,
            sum(case when date_part('day', 
                age(order_delivered_customer_date, 
                    order_purchase_timestamp)) BETWEEN 6 and 14 then 1 
                end) * 1.0 / count(order_id) * 100 as in_two_weeks,
            sum(case when date_part('day', 
                age(order_delivered_customer_date,
                order_purchase_timestamp)) > 14 then 1 
            end) * 1.0 / count(order_id) * 100 as more_than_two_weeks            
        FROM orders
        WHERE order_status = 'delivered'
        and date_part('day', age(order_delivered_customer_date, order_purchase_timestamp)) >= 0
        """

select(query)

under_two_days  in_one_week  in_two_weeks  more_than_two_weeks
0        6.136623    15.990463     54.298746            23.574168

Most orders were delivered in 2 weeks (54.29%).

### Shipping limit date

When an order is approved, Olist platform creates a deadline requiring the merchant to handle the order to Olist logistics partner before that date. 

In the past 30 days, how many merchants meet this requirement?

In [168]:
query = """
        SELECT c.num_sellers, count(b.seller_id) as num_meet_deadline
               ,count(b.seller_id) * 1.0 / c.num_sellers * 100 as percentage
        FROM 
        (
            SELECT a.seller_id
                ,meet_deadline * 1.0 / count(distinct order_id) * 100 as percentage
            FROM order_items
            JOIN 
            (
                SELECT oi.seller_id
                    ,count(distinct o.order_id) as meet_deadline
                FROM orders o 
                JOIN order_items oi on o.order_id = oi.order_id
                WHERE order_status = 'delivered'
                and date_part('month', order_purchase_timestamp) = 8
                and date_part('year', order_purchase_timestamp) = 2018
                and date_part('day', age(shipping_limit_date, order_delivered_carrier_date)) >= 0
                GROUP BY 1
            ) a
            on order_items.seller_id = a.seller_id
            GROUP BY a.seller_id, meet_deadline
        ) b
        CROSS JOIN 
        (
            SELECT count(distinct seller_id) as num_sellers
            FROM orders o 
            JOIN order_items oi on o.order_id = oi.order_id
            WHERE order_status = 'delivered'
            and date_part('month', order_purchase_timestamp) = 8
            and date_part('year', order_purchase_timestamp) = 2018
        ) c
        GROUP BY 1
        """

select(query)

num_sellers  num_meet_deadline  percentage
0         1261               1232   97.700238

We have 1261 merchants who sucessfully delivered orders to their customers in August 2018. 

97.70% of them meet the shipping deadline, but only 10.71% has 97% or more on-time delivery rate.